go/vt/topotools/vschema_ddl.go

```diff
@@ -0,0 +1,162 @@
+		if _, ok := ks.Vindexes[name]; ok {
+			return nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "vindex %s already exists in keyspace %s", name, ksName)
+		}
+
+		// Make sure the keyspace has the sharded bit set to true
+		// if this is the first vindex defined in the keyspace.
+		if len(ks.Vindexes) == 0 {
+			ks.Sharded = true
+		}
+
+		owner, params := ddl.VindexSpec.ParseParams()
+		ks.Vindexes[name] = &vschemapb.Vindex{
+			Type:   ddl.VindexSpec.Type.String(),
+			Params: params,
+			Owner:  owner,
+		}
+
+		return ks, nil
+
+	case sqlparser.AddColVindexStr:
+		// Support two cases:
+		//
+		// 1. The vindex type / params / owner are specified. If the
+		//    named vindex doesn't exist, create it. If it does exist,
+		//    require the parameters to match.
+		//
+		// 2. The vindex type is not specified. Make sure the vindex
+		//    already exists.
+		spec := ddl.VindexSpec
+		name := spec.Name.String()
+		if !spec.Type.IsEmpty() {
+			owner, params := spec.ParseParams()
+			if vindex, ok := ks.Vindexes[name]; ok {
+				if vindex.Type != spec.Type.String() {
+					return nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "vindex %s defined with type %s not %s", name, vindex.Type, spec.Type.String())
+				}
+				if vindex.Owner != owner {
+					return nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "vindex %s defined with owner %s not %s", name, vindex.Owner, owner)
+				}
+				if (len(vindex.Params) != 0 || len(params) != 0) && !reflect.DeepEqual(vindex.Params, params) {
+					return nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "vindex %s defined with different parameters", name)
+				}
+			} else {
+				// Make sure the keyspace has the sharded bit set to true
+				// if this is the first vindex defined in the keyspace.
+				if len(ks.Vindexes) == 0 {
+					ks.Sharded = true
+				}
+				ks.Vindexes[name] = &vschemapb.Vindex{
+					Type:   spec.Type.String(),
+					Params: params,
+					Owner:  owner,
+				}
+			}
+		} else {
+			if _, ok := ks.Vindexes[name]; !ok {
+				return nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "vindex %s does not exist in keyspace %s", name, ksName)
+			}
+		}
+
+		// If this is the first vindex being defined on the table, create
+		// the empty table record
+		if table == nil {
+			table = &vschemapb.Table{
+				ColumnVindexes: make([]*vschemapb.ColumnVindex, 0, 4),
+			}
+		}
+
+		// Make sure there isn't already a vindex with the same name on
+		// this table.
+		for _, vindex := range table.ColumnVindexes {
+			if vindex.Name == name {
+				return nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "vindex %s already defined on table %s", name, tableName)
+			}
+		}
+
+		columns := make([]string, len(ddl.VindexCols), len(ddl.VindexCols))
+		for i, col := range ddl.VindexCols {
+			columns[i] = col.String()
+		}
+		table.ColumnVindexes = append(table.ColumnVindexes, &vschemapb.ColumnVindex{
+			Name:    name,
+			Columns: columns,
+		})
+		ks.Tables[tableName] = table
+
+		return ks, nil
+
+	case sqlparser.DropColVindexStr:
```

Comment:
I think a nice feature here is to delete the vindex altogether if no other tables are using it.

Reply:
I think it's better to add an explicit syntax for dropping the vindex rather than implicit by dropping the last reference